src/backend/opencl/kernel/sparse_arith.hpp

```diff
@@ -272,5 +278,107 @@ namespace opencl
+            NDRange global(divup(M, local[0])*local[0], 1, 1);
+
+            nnzC = 0;
+            cl::Buffer* out = bufferAlloc(sizeof(unsigned));
+            getQueue().enqueueWriteBuffer(*out, CL_TRUE, 0, sizeof(unsigned), &nnzC);
+
+            calcNNZop(EnqueueArgs(getQueue(), global, local),
+                      *out, *outRowIdx.data, M,
+                      cl::Local(local[0]*sizeof(unsigned int)),
+                      *lrowIdx.data, *lcolIdx.data,
+                      *rrowIdx.data, *rcolIdx.data);
+            getQueue().enqueueReadBuffer(*out, CL_TRUE, 0, sizeof(unsigned), &nnzC);
+
+            CL_DEBUG_FINISH(getQueue());
+        }
+
+        template<typename T, af_op_t op>
+        void ssArithCSR(Param oVals, Param oColIdx,
+                const Param oRowIdx, const uint M, const uint N,
+                unsigned nnzA, const Param lVals, const Param lRowIdx, const Param lColIdx,
+                unsigned nnzB, const Param rVals, const Param rRowIdx, const Param rColIdx)
+        {
+            std::string refName = std::string("ss_arith_csr_") +
+                                  getOpString<op>() + "_" +
+                                  std::string(dtype_traits<T>::getName());
+            int device = getActiveDeviceId();
+            kc_entry_t entry = kernelCache(device, refName);
+
+            if (entry.prog==0 && entry.ker==0) {
+                ToNumStr<T> toNumStr;
+                std::ostringstream options;
+                options << " -D T="  << dtype_traits<T>::getName()
+                        << " -D OP=" << getOpString<op>()
+                        << " -D ZERO=(T)(" << scalar_to_option(scalar<T>(0)) << ")";
+
+                if((af_dtype) dtype_traits<T>::af_type == c32 ||
+                        (af_dtype) dtype_traits<T>::af_type == c64) {
+                    options << " -D IS_CPLX=1";
+                } else {
+                    options << " -D IS_CPLX=0";
+                }
+                if (std::is_same<T, double>::value ||
+                        std::is_same<T, cdouble>::value) {
+                    options << " -D USE_DOUBLE";
+                }
+
+                const char *kerStrs[] = { sparse_arith_common_cl, sp_sp_arith_csr_cl };
+                const int kerLens[] = { sparse_arith_common_cl_len, sp_sp_arith_csr_cl_len };
+
+                Program prog;
+                buildProgram(prog, 2, kerStrs, kerLens, options.str());
+                entry.prog = new Program(prog);
+                entry.ker = new Kernel(*entry.prog, "ssarith_csr_kernel");
+
+                addKernelToCache(device, refName, entry);
+            }
+            auto arithOp = KernelFunctor<Buffer, Buffer,
+                 Buffer, unsigned, unsigned,
+                 unsigned, const Buffer,
+                 const Buffer, const Buffer,
+                 unsigned, const Buffer,
+                 const Buffer, const Buffer>(*entry.ker);
+
+            NDRange local(32, 1);
```

Comment:
To reduce the total number of blocks/work-groups launched thereby allowing blocks(M-rows) to be much larger. 

Since, 32 is warp size in CUDA (probably, 64 for AMD GPUs), I chose 32 to avoid more than one warp/wavefront per block. It could be more than 32 I guess.